### PR TITLE
Add integer warning checks

### PR DESCRIPTION
**Pull request by cothan**
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

# Add clang warning checks

This PR will simpify or perhap make the CBMC proof tighter. 

Before I proceed to fix the warnings (later become error because we use `Werror`), let me know if this is a good idea. 

Please see the CI for error

### CODE REVIEW

**Comment by cothan** [comment]
The errors in ML-KEM are shown up in CI

**Comment by hanno-becker** [comment]
Hi @cothan, could you start working through the first set of warnings to give us a sense of the scope of changes?

**Comment by cothan** [comment]
@hanno-becker , sure. It's seem like we need to do manual casting in : https://github.com/pq-code-package/mlkem-c-aarch64/pull/248/commits/b7e25bdc67fe1f6488b6474b8766f444daa293a4

**Comment by cothan** [comment]
Thanks @rod-chapman for the comamds, I check locally with the commit: https://github.com/pq-code-package/mlkem-c-aarch64/pull/248/commits/96fabd933745f5c16472d2e0e0902abdab2f1890 
```
KYBER_K=2 ./run-cbmc-proofs.py --summarize -j8
KYBER_K=3 ./run-cbmc-proofs.py --summarize -j8
KYBER_K=4 ./run-cbmc-proofs.py --summarize -j8
```
are return success.

**Comment by cothan** [comment]
it's seem like the last part that I don't know how to cast is: 

```
mlkem/verify.c:68:9: warning: conversion to ‘int16_t’ {aka ‘short int’} from ‘int’ may change the sign of the result [-Wsign-conversion]
   68 |   *r ^= b & ((*r) ^ v);
      |         ^
```

in 

```c
void cmov_int16(int16_t *r, int16_t v, uint16_t b) {
// CBMC issues false alarms here for the implicit conversions between
// uint16_t and int, so disable "conversion-check" here for now.
// Revisit this when proof of this unit with contracts is attempted.
#pragma CPROVER check push
#pragma CPROVER check disable "conversion"
  b = -b;
  *r ^= b & ((*r) ^ v);
#pragma CPROVER check pop
}
```

File: mlkem/polyvec.c [hunk 1]
```diff
@@ -27,8 +27,7 @@ void polyvec_compress(uint8_t r[KYBER_POLYVECCOMPRESSEDBYTES],
   for (i = 0; i < KYBER_K; i++) {
     for (j = 0; j < KYBER_N / 8; j++) {
       for (k = 0; k < 8; k++) {
-        t[k] = a->vec[i].coeffs[8 * j + k];
-        t[k] += ((int16_t)t[k] >> 15) & KYBER_Q;
+        t[k] = scalar_signed_to_unsigned_q_16(a->vec[i].coeffs[8 * j + k]);
```
Comment (hanno-becker) [inline mlkem/polyvec.c:30]: Let's not mix functional changes with the changes needed to address the newly added warnings.

Comment (cothan) [reply mlkem/polyvec.c:30]: The original code convert signed to unsigned, I found the `scalar_signed_to_unsigned_q_16` is the perfect function to do so. 
Strange, in poly.c we have bound check and proof for `poly_compress`: 

```c
void poly_compress(uint8_t r[KYBER_POLYCOMPRESSEDBYTES], const poly *a) {
  POLY_UBOUND(a, KYBER_Q);
  uint8_t t[8] = {0};

  const int num_blocks = KYBER_N / 8;

#if (KYBER_POLYCOMPRESSEDBYTES == 128)
  for (int i = 0; i < num_blocks; i++)
      // clang-format off
        ASSIGNS(i, OBJECT_WHOLE(t), OBJECT_WHOLE(r))
        INVARIANT(i >= 0 && i <= num_blocks)
        DECREASES(num_blocks - i)
    // clang-format on
    {
      for (int j = 0; j < 8; j++).
...
```
However, the `polyvec_compress` isn't protected the same way as `poly_compress`, although it has division resistant code.

Comment (cothan) [reply mlkem/polyvec.c]: revert in https://github.com/pq-code-package/mlkem-c-aarch64/pull/248/commits/89089b6c7e90725904417001c5b17f2352ed6e1d

File: mlkem/polyvec.c [hunk 2]
```diff
@@ -58,8 +57,7 @@ void polyvec_compress(uint8_t r[KYBER_POLYVECCOMPRESSEDBYTES],
   for (i = 0; i < KYBER_K; i++) {
     for (j = 0; j < KYBER_N / 4; j++) {
       for (k = 0; k < 4; k++) {
-        t[k] = a->vec[i].coeffs[4 * j + k];
-        t[k] += ((int16_t)t[k] >> 15) & KYBER_Q;
+        t[k] = scalar_signed_to_unsigned_q_16(a->vec[i].coeffs[4 * j + k]);
```
[the remark this answers is not in the document]
Comment (cothan) [reply mlkem/polyvec.c]: revert in https://github.com/pq-code-package/mlkem-c-aarch64/pull/248/commits/89089b6c7e90725904417001c5b17f2352ed6e1d

**Comment by hanno-becker** [comment]
@cothan Perhaps wait a bit, I'm not convinced yet we want this change, and I don't want you to invest too much.

**Comment by rod-chapman** [comment]
I think we should stick to warning switches that are common to both recent versions of gcc and clang.

**Pull request closed without merging**